core/chains/evm/gas/arbitrum_estimator.go

```diff
@@ -0,0 +1,199 @@
+package gas
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+	"golang.org/x/exp/slices"
+
+	evmclient "github.com/smartcontractkit/chainlink/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+type ArbConfig interface {
+	EvmGasLimitMax() uint32
+}
+
+type ethClient interface {
+	CallContract(ctx context.Context, msg ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+}
+
+// arbitrumEstimator is an Estimator which extends l2SuggestedPriceEstimator to use getPricesInArbGas() for gas limit estimation.
+type arbitrumEstimator struct {
+	utils.StartStopOnce
+
+	cfg ArbConfig
+
+	Estimator // *l2SuggestedPriceEstimator
+
+	client     ethClient
+	pollPeriod time.Duration
+	logger     logger.Logger
+
+	getPricesInArbGasMu sync.RWMutex
+	perL2Tx             uint32
+	perL1CalldataUnit   uint32
+
+	chForceRefetch chan (chan struct{})
+	chInitialised  chan struct{}
+	chStop         chan struct{}
+	chDone         chan struct{}
+}
+
+func NewArbitrumEstimator(lggr logger.Logger, cfg ArbConfig, rpcClient rpcClient, ethClient ethClient) Estimator {
+	lggr = lggr.Named("ArbitrumEstimator")
+	return &arbitrumEstimator{
+		cfg:            cfg,
+		Estimator:      NewL2SuggestedPriceEstimator(lggr, rpcClient),
+		client:         ethClient,
+		pollPeriod:     10 * time.Second,
+		logger:         lggr,
+		chForceRefetch: make(chan (chan struct{})),
+		chInitialised:  make(chan struct{}),
+		chStop:         make(chan struct{}),
+		chDone:         make(chan struct{}),
+	}
+}
+
+func (a *arbitrumEstimator) Start(ctx context.Context) error {
+	return a.StartOnce("ArbitrumEstimator", func() error {
+		if err := a.Estimator.Start(ctx); err != nil {
+			return errors.Wrap(err, "failed to start gas price estimator")
+		}
+		go a.run()
+		<-a.chInitialised
+		return nil
+	})
+}
+func (a *arbitrumEstimator) Close() error {
+	return a.StopOnce("ArbitrumEstimator", func() (err error) {
+		close(a.chStop)
+		err = errors.Wrap(a.Estimator.Close(), "failed to stop gas price estimator")
+		<-a.chDone
+		return
+	})
+}
+
+// GetLegacyGas estimates both the gas price and the gas limit.
+func (a *arbitrumEstimator) GetLegacyGas(calldata []byte, l2GasLimit uint32, maxGasPriceWei *big.Int, opts ...Opt) (gasPrice *big.Int, chainSpecificGasLimit uint32, err error) {
```

Comment:
Would be nice to pass a context to all functions that can make a network request